web/src/stream_popover.js

```diff
@@ -207,7 +209,107 @@ function build_stream_popover(opts) {
+            current_message_id,
+            (is_oldest, is_newest) => {
+                if (is_oldest) {
+                    resolve("first");
+                } else if (is_newest) {
+                    resolve("last");
+                } else {
+                    resolve("intermediate");
+                }
+            },
+        );
+    });
+}
+
+async function get_message_placement_in_conversation(
+    current_stream_id,
+    topic_name,
+    current_message_id,
+) {
+    // First we check if the placement of the message can be determined
+    // in the current message list. This allows us to avoid a server call
+    // in most cases.
+
+    // We check if the view contains the entire conversation of which
+    // the message is a part of. The #design stream at CZO is an example of
+    // such a view.
+    if (current.data.filter.supports_collapsing_recipients()) {
+        // Next we check if we are in a conversation view, examples
+        // of such views are the topics and the DM view. If we are
+        // in a conversation view, we check if the message is the
+        // first or the last message in the current view. If not, we
+        // can conclude that the message is an intermediate message.
+        if (current.data.filter.is_conversation_view()) {
+            if (
+                current.data.fetch_status.has_found_oldest() &&
+                current.data.first().id === current_message_id
+            ) {
+                return "first";
+            } else if (
+                current.data.fetch_status.has_found_newest() &&
+                current.data.last().id === current_message_id
+            ) {
+                return "last";
+            }
+            return "intermediate";
+        }
+
+        // If we are not in a conversation view, but still know
+        // the view contains the entire conversation, we check if
+        // we can find the adjacent messages in the current view
+        // through which we can determine if the message is an
+        // intermediate message or not.
+        const msg_list = current.data.all_messages();
+        let is_next_found = false;
+        let is_prev_found = false;
+        const current_dict = {
+            stream_id: current_stream_id,
+            topic: topic_name,
+        };
+
+        for (let i = msg_list.length - 1; i >= 0; i -= 1) {
+            const message = msg_list[i];
+            const message_dict = {
+                stream_id: message.stream_id,
+                topic: message.topic,
+            };
+
+            if (util.same_stream_and_topic(current_dict, message_dict)) {
+                if (message.id > current_message_id) {
+                    is_next_found = true;
+                } else if (message.id < current_message_id) {
+                    is_prev_found = true;
+                }
+
+                if (is_next_found && is_prev_found) {
+                    return "intermediate";
+                }
+            }
+        }
```

Comment:
We can tell whether it's "last" or "first" much of the time here too; added conditions for that.